File: C++/Different-Ways-to-Add-Parentheses.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string input) {
        vector<int> res;
        int size = input.size();
        for (int i = 0; i < size; i ++) {
        	if (input[i] == '+' || input[i] == '-' || input[i] == '*') {
        		vector<int> left = diffWaysToCompute(input.substr(0, i));
        		vector<int> right = diffWaysToCompute(input.substr(i + 1));
        		for (auto n1 : left) {
        			for (auto n2 : right) {
        				if (input[i] == '+') {
        					res.push_back(n1 + n2);
        				} else if (input[i] == '-') {
        					res.push_back(n1 - n2);
        				} else {
        					res.push_back(n1 * n2);
        				}
        			}
        		}
        	}
        }
        if (res.empty()) {
        	res.push_back(atoi(input.c_str()));
        }
        return res;
    }
};

File: Java/Different-Ways-to-Add-Parentheses.java
public class Solution {
    public List<Integer> diffWaysToCompute(String input) {
        List<Integer> res = new ArrayList<Integer>();
        for (int i = 0; i < input.length(); i ++) {
            if (!Character.isDigit(input.charAt(i))) {
                List<Integer> left = diffWaysToCompute(input.substring(0, i));
                List<Integer> right = diffWaysToCompute(input.substring(i + 1));
                for (int j = 0; j < left.size(); j ++) {
                    for (int k = 0; k < right.size(); k ++) {
                        char ch = input.charAt(i);
                        if (ch == '+') res.add(left.get(j) + right.get(k));
                        else if (ch == '-') res.add(left.get(j) - right.get(k));
                        else res.add(left.get(j) * right.get(k));
                    }
                }
            }
        }
        if (res.isEmpty()) res.add(Integer.parseInt(input));
        return res;
    }
}